### DISCUSSION TITLE
Change json serialization strategy

### DISCUSSION THREAD

**Discussion by matheus1lva** [prisma/prisma#18931] [Q&A]
For json database column prisma uses plain `JSON.parse` in order to make the result readable, but there are some items that are not really well serializable like bigInt values stored inside those `JSON` objects. I would like to be able to parse them correctly to `BigInt`. Is there a way to override the deserialization for those fields?

**Comment by ludralph** [comment]
Hi @matheus1lva 👋 

Can you please give an example that elaborates on what you're referring to?

**Reply by matheus1lva** [reply]
Table in table X i have a json column where the value store contains BigInt values, in the example, amount.
```
{"amount": 12341234123412341234123412341234 }
```

As it gets to javascript land, it gets squased to float right? 

```
{amount: 1.2341234123412342e+31}
```

I wanna be able to convert them to use BigInt constructor when it arrives to userland, or string if that is the case.

**Reply by ludralph** [reply]
HI @matheus1lva, JSON does not support BigInt type. As a workaround, we will need to write a function that converts BigInt to String.
```js
const convertBigIntToString = (obj) => {
 return JSON.stringify(obj, (key, value) => {
  if (typeof value === 'bigint') {
      return value.toString() + 'n';
  } 
      return value;

});
```
Using this sample schema
```prisma
model Order {
  id Int @id @default(autoincrement())
  amount Float
  items Json
}
```
We can then have a query
```js
const res = await prisma.order.create({
  data: {
    amount: 500.30,
    items: convertBigIntToString({"amount": 12341234123412341234123412341234n })
  }
})
```
The output will be of the form 
```js
{
  id: 4,
  amount: 500.3,
  items: '{"amount":"12341234123412341234123412341234n"}'
}
```
To parse the JSON, we can pass a function as the second argument (reviver) to the JSON.parse() method.

```js
const convertJSONToBigInt = (j) => {
  return JSON.parse(j, (key, value) => {
    if (typeof value === 'string' && /^\d+n$/.test(value)) {
      return BigInt(value.slice(0, -1));
    }
    return value;
  });
}
```
Calling this function 

```js
convertJSONToBigInt(res.items)
```
returns the output

```js
{ amount: 12341234123412341234123412341234n }
```

Do let me know if you have further questions

**Reply by matheus1lva** [reply]
What if i already have that data onto db and want to use on prisma? Can i create a converter that does the opposite? From db to js?

**Reply by ludralph** [reply]
That should be possible i believe

**Reply by matheus1lva** [reply]
whenever i call `prismaClient.someEntity.findFirst({...});` it will return the data already processed by prisma. How do i plug this custom converter in between this process?

**Reply by ludralph** [reply]
Do you would want to pull the data from the database and convert the amount field to JSON?

**Reply by matheus1lva** [reply]
no

Given this table schema:

| - | Type  |
|---|---|
| A  | string  |
| B  | JSON  |


That json column has this value:
```
{amount: BIGINT, somethingElse: string}
```

As you see, when prisma pull this table, it does a JSON.parse(table.B), which converts the value of amount from BIGINT to float.

I want when prisma do the deserialization it maintains the BIGINT value.

**Reply by ludralph** [reply]
Hi @matheus1lva, I don't think that is possible at the moment because of the non-support of Bigint type in JSON.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!

**Comment by eyueldk** [comment]
This is such a useful feature. I wanted to serialize/deserialize some json fields using superjson.
